File: explore/medium/MajorityElement.js
/**
 * Given an array of size n, find majority element which appears more then n/2 times. A majority element always exists.
 * I would use a count hash and iterate through the nums array, but I would check intermittently.
 * 
 * @param {number[]} nums
 * @return {number}
 */
const majorityElement = function(nums) {
	let countHash = {};

	nums.forEach((num) => {
		countHash[num] ? countHash[num]++ : (countHash[num] = 1);
	});

	for (let key in countHash) {
		if (countHash[key] > nums.length / 2) return key;
	}
};

File: golang/majority_element.go
// given an array of nums, return the majority element.
func majorityElement(nums []int) int {
	freq := map[int]int{}
	for _, num := range nums {
		freq[num]++
	}

	for num, count := range freq {
		if count > len(nums)/2 {
			return num
		}
	}

	return -1
}

func majorityElement(nums []int) int {
	major := nums[0]
	count := 1

	for i := 1; i < len(nums); i++ {
		if count == 0 {
			major = nums[i]
			count = 1
		} else if nums[i] == major {
			count++
		} else {
			count--
		}
	}

	return major
}